### PR TITLE
Deprecate salt.payload.Serial

### LINKED ISSUE #60953: [TECH DEBT] Deprecate salt.payload.Serial
### Description of the tech debt to be addressed, include links and screenshots

`salt.payload.Serial` is un-needed. We only use msgpack for serialization. Deprecate `salt.payload.Serial` and just use the `loads` and `dumps` methods directly on `salt.payload`.

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

- Deprecate `salt.payload.Serial`
- Move `salt.payload` tests to pytest.
- fixes: #60953

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### CODE REVIEW

File: salt/payload.py [hunk 1]
```diff
@@ -58,186 +58,194 @@ def format_payload(enc, **kwargs):
...
+class Serial:
+    """
+    Create a serialization object, this object manages all message
+    serialization in Salt
+    """
+
+    def __init__(self, *args, **kwargs):
+        salt.utils.versions.warn_until(
+            "Chlorine",
+            "The `salt.payload.Serial` class has been depricated, "
```
Comment (s0undt3ch) [inline salt/payload.py:242]:
```suggestion
            "The `salt.payload.Serial` class has been deprecated, "
```

File: salt/payload.py [hunk 2]
```diff
@@ -58,186 +58,194 @@ def format_payload(enc, **kwargs):
...
+            "please use `salt.payload.loads` and `salt.payload.dumps`"
```
Comment (s0undt3ch) [inline salt/payload.py]:
```suggestion
            "and is set to be removed in {version}. "
            "Please use `salt.payload.loads` and `salt.payload.dumps`."
```
?

File: tests/pytests/functional/test_payload.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        Optional "sleep" can be sent to delay response
+        """
+        context = zmq.Context()
+        socket = context.socket(zmq.REP)
+        socket.bind("tcp://*:{}".format(port))
+        while event.is_set():
+            try:
+                #  Wait for next request from client
+                message = socket.recv(zmq.NOBLOCK)
+                print("GOT MSG {!r}".format(message))
```
Comment (s0undt3ch) [inline tests/pytests/functional/test_payload.py:50]:
```suggestion
```
Either log it or remove it.

File: tests/pytests/functional/test_payload.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    "Disabled until we can figure out how to make this more reliable."
+)
+def test_timeout(sreq, echo_server):
+    """
+    Test SREQ Timeouts
+    """
+    # client-side timeout
+    start = time.time()
+    # This is a try/except instead of an assertRaises because of a possible
+    # subtle bug in zmq wherein a timeout=0 actually exceutes a single poll
```
Comment (s0undt3ch) [inline tests/pytests/functional/test_payload.py:108]:
```suggestion
    # subtle bug in zmq wherein a timeout=0 actually executes a single poll
```

File: tests/pytests/functional/test_payload.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    try:
+        sreq.send("clear", "foo", tries=0, timeout=0)
+    except salt.exceptions.SaltReqTimeoutError:
+        pass
+    assert time.time() - start < 1  # ensure we didn't wait
+
+    # server-side timeout
+    log.info("Sending tries=1, timeout=1")
+    start = time.time()
+    with self.assertRaises(salt.exceptions.SaltReqTimeoutError):
```
Comment (s0undt3ch) [inline tests/pytests/functional/test_payload.py]:
```suggestion
    with pytest.raises(salt.exceptions.SaltReqTimeoutError):
```

File: tests/pytests/functional/test_payload.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        sreq.send("clear", {"sleep": 2}, tries=1, timeout=1)
+    assert time.time() - start >= 1  # ensure we actually tried once (1s)
+
+    # server-side timeout with retries
+    log.info("Sending tries=2, timeout=1")
+    start = time.time()
+    with self.assertRaises(salt.exceptions.SaltReqTimeoutError):
```
Comment (s0undt3ch) [inline tests/pytests/functional/test_payload.py]:
```suggestion
    with pytest.raises(salt.exceptions.SaltReqTimeoutError):
```

File: tests/pytests/unit/cache/test_mysql_cache.py [hunk 7]
```diff
@@ -49,14 +49,12 @@ def test_store(master_config):
...
 
-    serializer = salt.payload.Serial(master_config)
-
     mock_connect_client = MagicMock()
     with patch.object(mysql_cache, "_init_client") as mock_init_client:
         with patch.dict(
             mysql_cache.__context__,
             {
-                "serial": serializer,
+                "serial": salt.payload,
```
Comment (s0undt3ch) [inline tests/pytests/unit/cache/test_mysql_cache.py:57]: Could we possibly remove the `serial` key and update the module?

File: tests/pytests/unit/test_payload.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+import salt.exceptions
+import salt.payload
+from salt.utils import immutabletypes
+from salt.utils.odict import OrderedDict
+from tests.support.unit import TestCase, skipIf
+
+log = logging.getLogger(__name__)
+
+
+def assertNoOrderedDict(data):
```
Comment (s0undt3ch) [inline tests/pytests/unit/test_payload.py:18]:
```suggestion
def assert_no_ordered_dict(data):
```

File: tests/pytests/unit/test_payload.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        == b"\x81\xc7\x18N20010203T04:05:06.000007\xc7\x18N20010203T04:05:06.000007"
+    )
+    assert idata == odata
+
+
+def test_verylong_dump_load():
+    """
+    Test verylong encoder/decoder
+    """
+    payload = salt.payload.Serial("msgpack")
```
Comment (s0undt3ch) [inline tests/pytests/unit/test_payload.py:55]:
```suggestion
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Additionally, all `self.serial = salt.payload` code could/should be removed and any references to `self.serial` updated properly.

**Comment by dwoz** [comment]
> Additionally, all `self.serial = salt.payload` code could/should be removed and any references to `self.serial` updated properly.

fixed

File: tests/unit/cache/test_localfs.py [hunk 10]
```diff
@@ -102,7 +102,7 @@ def test_store_success(self):
         tmp_dir = tempfile.mkdtemp(dir=RUNTIME_VARS.TMP)
 
         # Use the helper function to create the cache file using localfs.store()
-        self._create_tmp_cache_file(tmp_dir, salt.payload.Serial(self))
+        self._create_tmp_cache_file(tmp_dir, salt.payload)
```
Comment (s0undt3ch) [inline tests/unit/cache/test_localfs.py:105]: `_create_tmp_cache_file` can now call `salt.payload` directly, do we really need to pass it as an argument?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/unit/cache/test_localfs.py [hunk 11]
```diff
@@ -137,15 +132,12 @@ def test_fetch_success(self):
...
-        serializer = salt.payload.Serial(self)
-
         # Use the helper function to create the cache file using localfs.store()
-        self._create_tmp_cache_file(tmp_dir, serializer)
+        self._create_tmp_cache_file(tmp_dir)
 
         # Now fetch the data from the new cache key file
         with patch.dict(localfs.__opts__, {"cachedir": tmp_dir}):
-            with patch.dict(localfs.__context__, {"serial": serializer}):
+            with patch.dict(localfs.__context__, {"serial": salt.payload}):
```
Comment (s0undt3ch) [inline tests/unit/cache/test_localfs.py:140]: Drop `__context__["serial"]` too from localfs?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/unit/cache/test_localfs.py (hunk 11 above)
Comment (dwoz) [reply tests/unit/cache/test_localfs.py:140]: Thought I tackled this already, maybe I missed this one?

Comment (s0undt3ch) [reply tests/unit/cache/test_localfs.py:140]: Probably a missed one.

**Review by s0undt3ch — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**